File: python/数组/1640. 能否连接形成数组.py
from typing import List
class Solution:
    def canFormArray(self, arr: List[int], pieces: List[List[int]]) -> bool:
        i = 0
        while i<len(arr):
            find = False
            for alist in pieces:
                j = 0
                if alist[0] == arr[i]:
                    find = True
                    while j < len(alist):
                        if arr[i] != alist[j]:
                            return False
                        j+=1
                        i+=1
                    break
            if not find:
                return False
            if i == len(arr):
                return True

class Solution:
    def canFormArray(self, arr: List[int], pieces: List[List[int]]) -> bool:
        from collections import defaultdict
        d = defaultdict(list)
        for alist in pieces:
            d[alist[0]] = alist
        i = 0
        while i < len(arr):
            if arr[i] not in d:
                return False
            else:
                for j in d[arr[i]]:
                    if j != arr[i]:
                        return False
                    i+=1
        return i == len(arr)


s = Solution()
arr = [85]
pieces = [[85]]
print(s.canFormArray(arr,pieces))

File: go/general/1640. 能否连接形成数组.go
package main

func canFormArray(arr []int, pieces [][]int) bool {
	d := make(map[int][]int)
	for _, v := range pieces {
		d[v[0]] = v
	}
	for i := 0; i < len(arr); {
		if _, ok := d[arr[i]]; !ok {
			return false
		}
		for _, v := range d[arr[i]] {
			if arr[i] != v {
				return false
			}
			i += 1
		}
	}
	return true
}
